[ClangImporter] Deduplicate clang buffers representing the same file

hnrklssn opened the pull request:
Clang can end up with multiple file IDs representing the same file, but in different contexts. For example, files in the current module always have file IDs >0, while files in imported modules always have file IDs <-1. As we end up compiling multiple modules, the same file can be seen both as the "current" module, and a transitive import. We don't want multiple Swift buffer IDs for the same file, as it breaks things like -verify that compare buffer IDs.

rdar://162661286

hnrklssn committed:
[ClangImporter] Deduplicate clang buffers representing the same file

Clang can end up with multiple file IDs representing the same file, but
in different contexts. For example, file in the current module always
have file IDs >0, while files in imported modules always have file IDs
<-1. As we end up compiling multiple modules, the same file can be seen
both as the "current" module, and a transitive import. We don't want
multiple Swift buffer IDs for the same file, as it breaks things like
-verify that compare buffer IDs.

rdar://162661286

egorzhdan reviewed (approved):
I'm not very familiar with this logic, but the change seems reasonable!

beccadax reviewed (approved):
Huh, didn't know that. LGTM.

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented on lib/Basic/SourceLoc.cpp:
```diff
@@ -99,6 +99,7 @@ unsigned
 SourceManager::addNewSourceBuffer(std::unique_ptr<llvm::MemoryBuffer> Buffer) {
   assert(Buffer);
   StringRef BufIdentifier = Buffer->getBufferIdentifier();
+  assert(BufIdentIDMap.find(BufIdentifier) == BufIdentIDMap.end());
```
I removed this assert in the end. It was a bit optimistic, and the complete fix is a bit out of scope for this PR.

hnrklssn committed:
[ClangImporter] check if buffer represents file before dedup

Virtual files from different modules can differ despite having identical
names (e.g. `<module-imports>`. This fixes some diagnostics related test
failures when these virtual files were (incorrectly) deduplicated and
the module imports from a previous module were shown, by checking if the
buffer has a corresponding clang::FileEntry.

hnrklssn commented:
@swift-ci please smoke test

hnrklssn committed:
Remove assert that path exists on file system

This would trigger when path prefixes are remapped, as the buffer
identifier is the remapped path, not the path on the file system. There
is no easy way to recover the original path, so drop the assert.

hnrklssn commented:
@swift-ci please smoke test

hnrklssn commented:
@swift-ci please test macos platform
